### DISCUSSION TITLE
Calling onChange doesn't update the actual value in hook form

### DISCUSSION THREAD

**Discussion by ShivamJoker** [react-hook-form/react-hook-form#9837] [Q&A]
## I have this Select input inside `Controller`

```jsx
  <Controller
    name="next_bedroom_option"
    control={control}
    defaultValue="none"
    render={({ field: { onChange, value } }) => (
      <Form.Select
        aria-label="Select if you want to start from scratch or want to go with previously filled information"
        onChange={(e) => onChange(e.currentTarget.value)}
        value={value}
      >
        <option value="none">Select an option</option>
        <option value="scratch">Start from scratch</option>
        {bedrooms_list.map((val, idx) => {
          return val ? (
            <option key={idx} value={idx}>
              Bedroom {idx + 1}
            </option>
          ) : null;
        })}
      </Form.Select>
    )}
  />
```

This is how the hook looks like, I am passing `formMethods` as a prop in child components

```jsx
  const formMethods = useForm({
    mode: "onChange",
    shouldUnregister: false,
    defaultValues: {
      room_type: "shared_room",
      property_type: "house",
      total_bedrooms: 1,
      total_bathrooms: 1,
      bedrooms_for_rent: 1,
      bedrooms_list: [],
      // will be 0,1,2,3
      current_bedroom: 0,
      next_bedroom_option: "none",
    },
```

I have setup `watch` on all the fields like this:

```jsx

useEffect(() => {
  const subscription = formMethods.watch((value, { name, type }) => {
    console.log("value updated", name, type);
    console.log(formMethods.getValues());
  });
  return () => subscription.unsubscribe();
}, [formMethods.watch]);
```

When I select the option the change event gets fired but the field doesn't get updated for some reason.

In the screenshot you can see next bedroom option always stays to the default value:

![image](https://user-images.githubusercontent.com/23727670/215258495-29513f05-7f40-44b9-a0df-8cf5cb963348.png)

> BTW the value in the dev tools also gets updated but not in the main form state.

I am using  `"react-hook-form": "^7.42.1"`

@bluebill1049 if you could help me with this, it would be great. I am struggling with this form wizard since 4 days.
Thanks!

**Comment by bluebill1049** [comment]
could you do the following?

```tsx
<Form.Select
  aria-label="Select if you want to start from scratch or want to go with previously filled information"
  onChange={(e) => onChange(e.currentTarget.value)}
  value={value} // log this value and also double check what type of data shape this controlled component requires to have
>
```

**Reply by ShivamJoker** [reply]
From the React bootstrap docs 

value | string \| arrayOf \| number |   | controlled by: onChange, initial prop: defaultValueThe value attribute of underlying input
-- | -- | -- | --

https://react-bootstrap.github.io/forms/select/

The value is coming as expected:

![image](https://user-images.githubusercontent.com/23727670/215259338-ed3050d0-ec02-44fb-99c2-c3561e9a5b55.png)

I added log like this

```jsx
 render={({ field: { onChange, value } }) => {
                      console.log("val", value);
                      return (
                        <Form.Select
```

**Reply by bluebill1049** [reply]
can you build a controlled component without hook form and see if its' working?

**Reply by ShivamJoker** [reply]
I tried it right now, it seems to be working perfectly fine with `useState`.

I logged the values it was getting updated.

**Reply by bluebill1049** [reply]
I think your component is getting unmounted and mounted with your implementation. If you can reproduce it in a code sandbox i can take a look without your other dependency.

**Reply by ShivamJoker** [reply]
The problem is that it is not getting re-rendered which is should have.
I added `useEffect` log to check, it still didn't render on value changes.

I will try to make a repro.

Also should I try calling `setValue` instead of `onChange` ?

**Reply by ShivamJoker** [reply]
Yes, calling `setValue` seems to reRender it.

**Reply by bluebill1049** [reply]
> I added useEffect log to check, it still didn't render on value changes.

useEffect is not going to be called, controller only re-render within its render function.

> Yes, calling setValue seems to reRender it.

I am not sure this is the root cause, onChange should re-render the cotnroller's render prop.
